ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/parser/TestStatement.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.test.parser;
+
+import io.confluent.ksql.parser.KsqlParser.ParsedStatement;
+import io.confluent.ksql.parser.tree.AssertStatement;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * A single line of a test execution. A {@code TestStatement} can consist of
+ * either:
+ * <ul>
+ *   <li> a {@link ParsedStatement}, which will issue an update to the ksqlDB, or </li>
+ *   <li> an {@link AssertStatement}, which will ensure data is accurate, or </li>
+ *   <li> a {@link TestDirective}, which will ensure some metadata on the test </li>
+ * </ul>
+ */
+public final class TestStatement {
```

Comment:
A cleaner API might be to store and return `Optional`s, e.g. 

```java
public Optional<ParsedStatement> getEngineStatement() {...}
```

This does away with the need for the `has` family of methods, and means the compiler/IDE can check for access to the optional without a matching `isPresent` etc, which you don't get by cutting your own version.


Also, consider adding a more functional  method that can take `Consumer`s for each type in the union, simplifying code needing to handle this tri-state type.

```java
   void accept(
     final Consumer<ParsedStatement> sqlConsumer, 
     final Consumer<AssertStatement<?> assertConsumer,
     final Consumer<TestDirective> directiveConsumer
  );
```

Reply:
I had thought about this, but I think in practice this makes the code harder to read and understand. This is not a class with three optional fields, this is a "OneOf" class - if there were only three get statements, then the API suggests that any combination of them can be present.

I do like your `accept` suggestion, though, in which case I just won't expose the fields at all and the discussion becomes moot 😂

Reply:
Not sure I follow why you think `hasXXX` methods better impart that it's a  "OneOf" class that using `Optional`.  But if its moot, then let's move on.


hummm... I guess you could call them `hasOnlyXXXX` to make the union clear. :p